File: todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/statistics/view/StatisticsViews.java
/*
 * Copyright 2016, The Android Open Source Project
 * Copyright (c) 2017-2018 Spotify AB
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.example.android.architecture.blueprints.todoapp.statistics.view;

import android.content.res.Resources;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;
import com.example.android.architecture.blueprints.todoapp.R;
import com.example.android.architecture.blueprints.todoapp.statistics.domain.StatisticsEvent;
import com.example.android.architecture.blueprints.todoapp.statistics.domain.StatisticsState;
import com.spotify.mobius.Connectable;
import com.spotify.mobius.Connection;
import com.spotify.mobius.functions.Consumer;
import javax.annotation.Nonnull;

public class StatisticsViews implements Connectable<StatisticsState, StatisticsEvent> {

  private final View mRoot;
  private final TextView mStatisticsTV;

  public StatisticsViews(LayoutInflater inflater, ViewGroup parent) {
    mRoot = inflater.inflate(R.layout.statistics_frag, parent, false);
    mStatisticsTV = mRoot.findViewById(R.id.statistics);
  }

  public View getRootView() {
    return mRoot;
  }

  @Nonnull
  @Override
  public Connection<StatisticsState> connect(Consumer<StatisticsEvent> output) {
    return new Connection<StatisticsState>() {
      @Override
      public void accept(StatisticsState state) {
        renderState(state);
      }

      @Override
      public void dispose() {}
    };
  }

  private void renderState(StatisticsState state) {
    state.match(
        loading -> mStatisticsTV.setText(R.string.loading),
        loaded -> {
          int numberOfCompletedTasks = loaded.completedCount();
          int numberOfIncompleteTasks = loaded.activeCount();
          if (numberOfCompletedTasks == 0 && numberOfIncompleteTasks == 0) {
            mStatisticsTV.setText(R.string.statistics_no_tasks);
          } else {
            Resources resources = mRoot.getContext().getResources();
            String activeTasksString =
                resources.getString(R.string.statistics_active_tasks, numberOfIncompleteTasks);
            String completedTasksString =
                resources.getString(R.string.statistics_completed_tasks, numberOfCompletedTasks);
            String displayString = activeTasksString + "\n" + completedTasksString;
            mStatisticsTV.setText(displayString);
          }
        },
        failed -> mStatisticsTV.setText(R.string.statistics_error));
  }
}

File: todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/statistics/ui/StatisticsViews.kt
/*
 * Copyright 2016, The Android Open Source Project
 * Copyright (c) 2017-2018 Spotify AB
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.example.android.architecture.blueprints.todoapp.statistics.ui

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import com.example.android.architecture.blueprints.todoapp.R
import com.example.android.architecture.blueprints.todoapp.statistics.StatisticsEvent
import com.example.android.architecture.blueprints.todoapp.statistics.StatisticsState
import com.example.android.architecture.blueprints.todoapp.util.onAccept
import com.spotify.mobius.Connectable
import com.spotify.mobius.Connection
import com.spotify.mobius.functions.Consumer

class StatisticsViews(inflater: LayoutInflater, parent: ViewGroup) : Connectable<StatisticsState, StatisticsEvent> {

    val rootView: View = inflater.inflate(R.layout.statistics_frag, parent, false)

    private val statisticsTV: TextView

    init {
        statisticsTV = rootView.findViewById(R.id.statistics)
    }

    override fun connect(output: Consumer<StatisticsEvent>): Connection<StatisticsState> {
        return onAccept<StatisticsState> { renderState(it) }.onDispose { }
    }

    private fun renderState(state: StatisticsState) {
        when (state) {
            is StatisticsState.Loading -> statisticsTV.setText(R.string.loading)
            is StatisticsState.Loaded -> {
                if (state.hasTasks()) {
                    statisticsTV.text = rootView.resources.getString(R.string.statistics_tasks_count, state.activeCount, state.completedCount)
                } else {
                    statisticsTV.setText(R.string.statistics_no_tasks)
                }
            }
            is StatisticsState.Failed -> statisticsTV.setText(R.string.statistics_error)
        }
    }

    private fun StatisticsState.Loaded.hasTasks() = this.activeCount != 0 || this.completedCount != 0
}
